### PR TITLE
[feature] Allow loading TLS certs from disk

### PR DESCRIPTION
**Pull request by daenney**
# Description

> If this is a code change, please include a summary of what you've coded, and link to the issue(s) it closes/implements.
>
> If this is a documentation change, please briefly describe what you've changed and why.

Currently, GtS only supports using the built-in LE client directly for TLS. However, admins may still want to use GtS directly (so without a reverse proxy) but with certificates provided through some other mechanism. They may have some centralised way of provisioning these things themselves, or simply prefer to use LE but with a different challenge like DNS-01 which is not supported by autocert.

This adds support for loading a public/private keypair from disk instead of using LE and reconfigures the server to use a TLS listener if we succeed in doing so.

Additionally, being able to load TLS keypair from disk opens up the path to using a custom CA for testing purposes avoinding the need for a constellation of containers and something like Pebble or Step CA to provide LE APIs.

## Checklist

Please put an x inside each checkbox to indicate that you've read and followed it: `[ ]` -> `[x]`

If this is a documentation change, only the first checkbox must be filled (you can delete the others if you want).

- [x] I/we have read the [GoToSocial contribution guidelines](https://github.com/superseriousbusiness/gotosocial/blob/main/CONTRIBUTING.md).
- [ ] I/we have discussed the proposed changes already, either in an issue on the repository, or in the Matrix chat.
- [x] I/we have performed a self-review of added code.
- [x] I/we have written code that is legible and maintainable by others.
- [x] I/we have commented the added code, particularly in hard-to-understand areas.
- [x] I/we have made any necessary changes to documentation.
- [ ] I/we have added tests that cover new code.
- [x] I/we have run tests and they pass locally with the changes.
- [x] I/we have run `go fmt ./...` and `golangci-lint run`.

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -418,6 +419,20 @@ In [./test/envparsing.sh](./test/envparsing.sh) there's a test for making sure t
...
 
+#### Federation
+
+By using the support for loading TLS files from disk it is possible to have two local instances with TLS to allow for (manually) testing federation.
+
+You'll need to set the following configuration options:
+* `GTS_TLS_CERTIFICATE_CHAIN`: poiting to a PEM-encoded certificate chain including the public certificate
+* `GTS_TLS_CERTIFICATE_KEY`: pointing to a PEM-encoded private key
+
+Additionally, for the Go HTTP client to recognise certificates issued by a custom CA as valid, you'll need to set one of:
```
Comment (daenney) [inline CONTRIBUTING.md:430]: Except on macOS or Windows: https://pkg.go.dev/crypto/x509#SystemCertPool

**Comment by daenney** [comment]
Go has built-in support for generating a CA or certificates for you. Expanding a bit on that it should be fairly trivial to have a test helper script to generate the CA + issue certificates for a pair of named instances.

```
go run /usr/lib/go/src/crypto/tls/generate_cert.go -help
Usage of /run/user/1000/go-build2063357202/b001/exe/generate_cert:
  -ca
    	whether this cert should be its own Certificate Authority
  -duration duration
    	Duration that certificate is valid for (default 8760h0m0s)
  -ecdsa-curve string
    	ECDSA curve to use to generate a key. Valid values are P224, P256 (recommended), P384, P521
  -ed25519
    	Generate an Ed25519 key
  -host string
    	Comma-separated hostnames and IPs to generate a certificate for
  -rsa-bits int
    	Size of RSA key to generate. Ignored if --ecdsa-curve is set (default 2048)
  -start-date string
    	Creation date formatted as Jan 1 15:04:05 2011
```

**Comment by NyaaaWhatsUpDoc** [comment]
> Go has built-in support for generating a CA or certificates for you. Expanding a bit on that it should be fairly trivial to have a test helper script to generate the CA + issue certificates for a pair of named instances. Or leverage something like [mkCert](https://github.com/FiloSottile/mkcert).
> 
> ```
> go run /usr/lib/go/src/crypto/tls/generate_cert.go -help
> Usage of /run/user/1000/go-build2063357202/b001/exe/generate_cert:
>   -ca
>     	whether this cert should be its own Certificate Authority
>   -duration duration
>     	Duration that certificate is valid for (default 8760h0m0s)
>   -ecdsa-curve string
>     	ECDSA curve to use to generate a key. Valid values are P224, P256 (recommended), P384, P521
>   -ed25519
>     	Generate an Ed25519 key
>   -host string
>     	Comma-separated hostnames and IPs to generate a certificate for
>   -rsa-bits int
>     	Size of RSA key to generate. Ignored if --ecdsa-curve is set (default 2048)
>   -start-date string
>     	Creation date formatted as Jan 1 15:04:05 2011
> ```

holy shit i can't believe i never knew this

File: CONTRIBUTING.md (hunk 1 above)
Comment (NyaaaWhatsUpDoc) [reply CONTRIBUTING.md:430]: Fortunately we don't support either of these :p

**Comment by NyaaaWhatsUpDoc** [comment]
nice work :)

**Comment by tsmethurst** [comment]
Thanks for this!

**Pull request merged**